compiler/luci/pass/src/FoldGatherPass.cpp

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FoldGatherPass.h"
+#include "CircleOptimizerUtils.h"
+
+#include <luci/IR/CircleNodes.h>
+
+namespace
+{
+
+/**
+ * Fold to const if
+ *
+ * 1. params: const and dtype = S32 or S64
+ * 2. indices: const and dtype = S32 or S64
+ *
+ **/
+template <loco::DataType InputT, loco::DataType IndexT>
+bool fold_gather(luci::CircleGather *gather_node)
+{
+  const auto params = dynamic_cast<luci::CircleConst *>(gather_node->params());
+  const auto indices = dynamic_cast<luci::CircleConst *>(gather_node->indices());
+
+  const auto axis = gather_node->axis();
+  assert(axis >= 0);
+  if (axis < 0)
+    return false;
+
+  const auto name = gather_node->name();
+  assert(name.length() > 0);
+
+  auto constant = gather_node->graph()->nodes()->create<luci::CircleConst>();
+  constant->dtype(InputT);
+  constant->name(name + "_G");
+
+  const auto rank = params->rank();
+  assert(axis < rank);
+
+  constant->rank(rank + indices->rank() - 1);
+
+  assert(constant->rank() > 0);
+
+  std::vector<uint32_t> shape;
+  for (uint32_t i = 0; i < rank; ++i)
+  {
+    if (i != axis)
+    {
+      const auto dim = params->dim(i).value();
+      assert(dim >= 0 && dim <= std::numeric_limits<uint32_t>::max());
+      shape.push_back(dim);
+    }
+    else
+    {
+      for (uint32_t j = 0; j < indices->rank(); ++j)
+      {
+        const auto dim = indices->dim(j).value();
+        assert(dim >= 0 && dim <= std::numeric_limits<uint32_t>::max());
+        shape.push_back(dim);
+      }
+    }
+  }
+
+  uint32_t size = 1;
+  for (uint32_t i = 0; i < shape.size(); ++i)
+  {
+    constant->dim(i).set(shape.at(i));
+    size *= shape.at(i);
+  }
+
+  constant->size<InputT>(size);
+
+  uint32_t outer_size = 1;
```

Comment:
Why not using tflite kernel, which is well-tested ? You can see how to use tflite kernel for constant folding in `FoldDepthwiseConv2DPass`.

Reply:
I have tried to use tflite kernel here but I have some trouble with it. Kernel `../internal/reference/gather.h`  has `ruy::profiler::ScopeLabel label("Gather");` and this pass doesn't building with it, due to it doesn't see 'ruy` declaration.
How I can handle with it? Or I can keep current implementation (this implementation is the copy of tflite kernel implementation).

Reply:
I see. Let's go with the current implementation. Thanks.